draft-ietf-quic-recovery.md

```diff
@@ -162,43 +162,209 @@ latency before a userspace QUIC receiver processes a received packet.
+
+The RECOMMENDED value for kDelayedAckTimeout is 25ms.
+
+A PTO value of at least 2*SRTT ensures that the ACK is overdue. Using a PTO of
+exactly 1*SRTT may generate spurious probes, and 2*SRTT is simply the next
+integral value of RTT.
+
+(TODO: These values of 2 and 1.5 are a bit arbitrary. Reconsider these.)
+
+If the Retransmission Timeout (RTO, {{rto}}) is smaller than the computed PTO,
+then a PTO is scheduled for the smaller RTO period.
+
+To reduce latency, it is RECOMMENDED that the sender send two sequential TLP
+packets before setting an RTO alarm. When the first TLP packet is sent, the TLP
+alarm should be scheduled for a second TLP. On the second TLP packet, an RTO
+alarm should be scheduled {{rto}}.
+
+A TLP packet SHOULD carry new data when available. If new data is unavailable, a
+TLP packet MAY retransmit unacknowledged data to potentially reduce recovery
+time. Since a TLP packet is sent as a probe into the network prior to
+establishing any packet loss, prior unacknowledged packets SHOULD NOT be marked
+as lost.
+
+A TLP packet MUST NOT be blocked by the sender's congestion controller. The
+sender MUST however count these bytes as additional bytes in flight, since a TLP
+adds network load without establishing packet loss.
+
+A sender will commonly not know that a packet being sent is a tail packet.
+Consequently, a sender may have to arm or adjust the TLP alarm on every sent packet.
+
+### Retransmission Timeout {#rto}
+
+A Retransmission Timeout (RTO) alarm is the final backstop for loss
+detection. The algorithm used in QUIC is based on the RTO algorithm for TCP
+{{!RFC5681}} and is additionally resilient to spurious RTOs.
+
+When the last TLP packet is sent, an alarm is scheduled for the RTO time. When
+this alarm fires, the sender sends two Retransmission Timeout (RTO) packets, to
+evoke acknowledgements from the receiver, and restarts the RTO alarm.
+
+Similar to TCP {{!RFC6298}}, the RTO alarm duration is set based on the
+following conditions:
+
+* When the final TLP packet is sent, RTO is set to max(SRTT + 4*RTTVAR, minRTO)
+
+* When an RTO alarm fires, RTO is set to twice its current value.
+
+The sender typically incurs a high latency penalty when an RTO is encountered,
+and this penalty increases exponentially in subsequent consecutive RTOs. Sending
+a single RTO packet therefore makes the connection very sensitive to single
+packet loss. Sending two packets instead of one significantly increases
+resilience to packet drop in both directions, thus reducing the probability of
+consecutive RTOs.
+
+QUIC's RTO algorithm differs from TCP in that an RTO alarm event is not
+considered a strong enough signal of packet loss. An RTO alarm fires only when
+there's a prolonged period of network silence, which could be caused by a change
+in the underlying network RTT.
+
+When an ack is received after one or more consecutive RTO alarm events, any
+packets with lower packet numbers than those acked MUST be marked as lost.
```

Comment:
Does this apply even if the ACK doesn't ACK the RTO packets, i.e. was an earlier delayed ACK?

Reply:
Good point, I think it should not.  In that case, we should fall out of RTO mode and go back into standard ack based recovery.

Reply:
I've modified the text here, PTAL.